sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala

```diff
@@ -331,6 +332,268 @@ case class RoundCeil(child: Expression, scale: Expression)
+             |$input, ${_scale}))""".stripMargin
+          case FloatType if (_scale <= 0) =>
+            s"""org.apache.spark.sql.catalyst.expressions.TruncNumber.trunc(
+             |$input, ${_scale}).floatValue()""".stripMargin
+          case DoubleType if (_scale <= 0) =>
+            s"""org.apache.spark.sql.catalyst.expressions.TruncNumber.trunc(
+             |$input, ${_scale}).doubleValue()""".stripMargin
+          case DecimalType.Fixed(_, _) =>
+            s"""Decimal.apply(
+             |org.apache.spark.sql.catalyst.expressions.TruncNumber.trunc(
+             |${input}.toJavaBigDecimal(), ${_scale}))""".stripMargin
+          case _ => s"$input"
+        }
+      })
+
+  /**
+   * Returns the [[DataType]] of the result of evaluating this expression. It is invalid to query
+   * the dataType of an unresolved expression (i.e., when `resolved` == false).
+   */
+  override lazy val dataType: DataType = {
+    child.dataType match {
+      case DecimalType.Fixed(p, s) =>
+        val newPosition =
+          if (_scale > 0) {
+            if (_scale >= s) {
+              s
+            } else {
+              _scale
+            }
+          } else {
+            0
+          }
+        DecimalType(p - s + newPosition, newPosition)
+      case t => t
+    }
+  }
+
+  /**
+   * Called by default [[eval]] implementation. If subclass of BinaryExpression keep the default
+   * nullability, they can override this method to save null-check code. If we need full control
+   * of evaluation process, we should override [[eval]].
+   */
+  override protected def nullSafeEval(input1: Any, input2: Any): Any = {
+    dataType match {
+      case ByteType if (_scale <= 0) =>
+        TruncNumber.trunc(input1.asInstanceOf[Byte].toLong, _scale).toByte
+      case ShortType if (_scale <= 0) =>
+        TruncNumber.trunc(input1.asInstanceOf[Short].toLong, _scale).shortValue
+      case IntegerType if (_scale <= 0) =>
+        TruncNumber.trunc(input1.asInstanceOf[Int].toLong, _scale).intValue
+      case LongType if (_scale <= 0) =>
+        TruncNumber.trunc(input1.asInstanceOf[Long], _scale).longValue
+      case FloatType =>
+        TruncNumber.trunc(input1.asInstanceOf[Float], _scale).floatValue
+      case DoubleType =>
+        TruncNumber.trunc(input1.asInstanceOf[Double], _scale).doubleValue
+      case DecimalType.Fixed(p, s) =>
+        Decimal(TruncNumber.trunc(input1.asInstanceOf[Decimal].toJavaBigDecimal, _scale))
+      case _ => input1
+    }
+  }
+}
+
+object TruncNumber {
+  /**
+   * To truncate whole numbers ; byte, short, int, long types
+   */
+  def trunc(input: Long, position: Int): Long = {
+    if (position >= 0) {
+      input
+    } else {
+      // position is -ve, truncate the number by absolute value of position
+      // eg: input 123 , scale -2 , result 100
+      val pow = Math.pow(10, Math.abs(position)).toLong
+      (input / pow) * pow
+    }
+  }
+
+  /**
+   * To truncate double and float type
```

Comment:
please convert each comment to a full sentence with a trailing period?

Reply:
Done

Reply:
This change does not seem to be uploaded to GitHub; did you forget to push a commit?